src/transform/src/join_implementation.rs

```diff
@@ -402,42 +421,167 @@ impl JoinImplementation {
+                return Ok(());
+            }
+
+            // To have reached here, we must be in our first run of join planning.
+            //
+            // We plan a differential join first.
+            let (differential_query_plan, differential_new_arrangements) = differential::plan(
+                relation,
+                &input_mapper,
+                &available_arrangements,
+                &unique_keys,
+                &cardinalities,
+                &filters,
+            )
+            .expect("Failed to produce a differential join plan");
+
+            // Binary joins _must_ be differential. We won't plan a delta join.
+            if num_inputs <= 2 {
+                // if inputs.len() == 0 then something is very wrong.
+                soft_assert_or_log!(num_inputs != 0, "join with no inputs");
+                // if inputs.len() == 1:
+                // Single input joins are filters and should be planned as
+                // differential plans instead of delta queries. Because a
+                // a filter gets converted into a single input join only when
+                // there are existing arrangements, without this early return,
+                // filters will always be planned as delta queries.
+                // (ggevay: This is an old comment, and I'm not sure whether a single-input join
+                // could still actually occur. It is not happening in any of our slts currently.)
+                soft_assert_or_log!(
+                    num_inputs != 1,
+                    "join with only one input (should be filter)"
+                );
+                // if inputs.len() == 2:
+                // We decided to always plan this as a differential join for now, because the usual
+                // advantage of a Delta join avoiding intermediate arrangements doesn't apply.
+                // See more details here:
+                // https://github.com/MaterializeInc/materialize/pull/16099#issuecomment-1316857374
+                // https://github.com/MaterializeInc/materialize/pull/17708#discussion_r1112848747
+                *relation = differential_query_plan;
+
+                return Ok(());
+            }
+
+            // We are plannign a multiway join for the first time.
+            //
+            // We compare the delta and differential join plans.
+            //
+            // A delta query requires that, for every order, there is an arrangement for every
+            // input except for the starting one. Such queries are viable when:
+            //
+            //   (a) all the arrangements already exist, or
+            //   (b) both:
+            //       (i) we wouldn't create more arrangements than a differential join would
+            //       (ii) `enable_eager_delta_joins` is on
+            //
+            //
+            // A differential join of k relations requires k-2 arrangements of intermediate
+            // results (plus k arrangements of the inputs).
```

Comment:
Where is the `k-2` reflected in the code?

Reply:
So in a differential join, we need arrangements on the result of every join except the last one; in a k-way join, that means k-2 arrangements. It's not reflected in the decisions in `action`---it's implicitly reflected in the way `differential::plan` adds new arrangements.

Reply:
Oh, wait: it _should_ have been reflected on line 843 (`let new_arrangements = ...`), but the `-2` was missing. Thank you, good catch!!